### ISSUE TITLE
Invalid link with persistent parameter

### ISSUE THREAD

**Issue by olexap** [contributte/menu-control#43]
Found a bug.

Trying to use route with language parameter which is persistent parameter in presenter.
All links generated in project are okay but links in menu-control don't respect actual value of the parameter. Instead there are just default parameter value for route.

**Comment by foxycode** [comment]
Thanks for reporting. Can you please add example code how to induce this bug?

**Comment by olexap** [comment]
<?php

declare(strict_types=1);

namespace App\Presenters;

use Carrooi\Menu\UI\IMenuComponentFactory;
use Carrooi\Menu\UI\MenuComponent;
use Nette;
use Nette\ComponentModel\IComponent;


final class HomepagePresenter extends Nette\Application\UI\Presenter
{
    /**
     * @persistent
     */
    public $lang;

    private $menuFactory;


    public function injectBasePresenter(IMenuComponentFactory $menuFactory)
    {
        $this->menuFactory = $menuFactory;
    }


    protected function createComponentMenu(): MenuComponent
    {
        return $this->menuFactory->create('front');
    }

}


<?php

declare(strict_types=1);

namespace App\Router;

use Nette;
use Nette\Application\Routers\RouteList;


final class RouterFactory
{
	use Nette\StaticClass;

	public static function createRouter(): RouteList
	{
		$router = new RouteList;
		$router->addRoute('[!<lang=cz>/]<presenter>/<action>[/<id>]', 'Homepage:default');
		return $router;
	}
}


neon:

parameters:


application:
	errorPresenter: Error
	mapping:
		*: App\*Module\Presenters\*Presenter


session:
	expiration: 14 days


services:
	router: App\Router\RouterFactory::createRouter


extensions:
    menu: Carrooi\Menu\DI\MenuExtension

menu:

  front:
    items:

      Home:
        action: Homepage:default

      Test:
        action: Homepage:test






default.latte:


{* This is the welcome page, you can delete it *}

{block content}

	{control menu}

	{link Homepage:test}
	<br>
	{plink Homepage:test}

```

**Comment by olexap** [comment]
[web-project.tar.gz](https://github.com/contributte/menu-control/files/4204993/web-project.tar.gz)

**Comment by olexap** [comment]
I trying to write my own menu component and I end at the same problem when I trying to generate url myself (not using latte). It seems like problem is LinkGenerator in nette. It ignores parameters defined in route and use defaults instead.
I have no idea how to make it work.

**Comment by foxycode** [comment]
Yes, this will be probably the cause, because `menu-control` uses `LinkGenerator` which don't known about current state of presenter persistent parameters and thus uses default values.

Solution would be create your own implementation of  `ILinkGenerator` and use `Presenter` for link generation, but we get rid of `Presenter` earlier, because we had some problems I don't remember.

**Comment by olexap** [comment]
Yes. Nette LinkGenerator is wrong. I wandering that maybe will be possible to use RouteList and Request to make own LinkGenerator but I don't know how to access parameters from router :(
Dead end at the moment.

**Comment by patrickkusebauch** [comment]
For somebody, who stumble upon this as I did, here is a hacky workaround:
```php
use Carrooi\Menu\IMenuItem;
use Carrooi\Menu\LinkGenerator\ILinkGenerator;
use Nette\Application\Application;
use Nette\Application\UI\Presenter;

final class NetteLinkGenerator implements ILinkGenerator
{

    private Presenter $presenter;

    public function __construct(Application $application)
	{
		$this->presenter = $application->getPresenter();
	}


	public function link(IMenuItem $item): string
	{
        if(($action = $item->getAction()) !== null) {
            return $this->presenter->link($action, $item->getActionParameters());
        }

        if (($link = $item->getLink()) !== null) {
            return $link;
        }

        return '#';
	}

}
```

The caveat is that now it recognizes relative vs absolute links, so in your config file, all your links now have to start with a colon.

I can give you no warranty, I just tried this out and persistent parameters works, but I have no idea if something else isn't broken.

**Comment by foxycode** [comment]
Since this is not `menu-cottrol` bug, closing issue.

**Comment by patrickkusebauch** [comment]
@foxycode I guess this really depends on the point of view. Since it is a presented and Nette component, but it does not work fully with Nette. It should be at least mentioned somewhere in the documentation, WDYT?

**Comment by foxycode** [comment]
@patrickkusebauch You're right. Will think about it.
